### PR TITLE
Runtime attachment feature

### PR DESCRIPTION
**Pull request by jeanbisutti**
Runtime attachment feature for _Application Insights._

The feature delegates attachment to [_Otel runtime attachment_ project](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/runtime-attach) and allows to manage things specific to _Application Insights_ (_applicationinsights.json_ and _version tracking_).

[A PR](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/354) was also  created to improve the _Otel runtime attachment_ (more checks and logs).

### CODE REVIEW

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java:264]: can you do runtime attach check at the beginning of the method and then return early?

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (heyams) [inline agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:57]: can you explain why line 45 and then setting it to false here?

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (heyams) [inline agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:61]: I saw you have another pr on opentelemetry. should you wait till that PR is merged and use the upstream to check if agent is attached?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java
Comment (jeanbisutti) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java:264]: @heyams Could you please explain why? If we return early at the beginning of the method then the os information won't be available for tracking.

Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java:264]: something like this: code between 
"// start runtime attach prefix" and 
"// end runtime attach prefix"

your change will end up with "ara_" or "fra_" depends on which rp integration it was. 
```
@Nullable
  private static String getCodelessSdkNamePrefix() {
    if (!DiagnosticsHelper.isRpIntegration()) {
      return null;
    }
    StringBuilder sdkNamePrefix = new StringBuilder(4);
    // start runtime attach prefix
    if (isRuntimeAttached()) {
      sdkNamePrefix.append("ra_");
      return;
    } // end runtime attach prefix
    sdkNamePrefix.append(DiagnosticsHelper.rpIntegrationChar());
    if (SystemInformation.isWindows()) {
      sdkNamePrefix.append("w");
    } else if (SystemInformation.isLinux()) {
      sdkNamePrefix.append("l");
    } else {
      LoggerFactory.getLogger("com.microsoft.applicationinsights.agent")
          .warn("could not detect os: {}", System.getProperty("os.name"));
      sdkNamePrefix.append("u");
    }
    sdkNamePrefix.append("r_"); // "r" is for "recommended"
    return sdkNamePrefix.toString();
  }
```

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (jeanbisutti) [reply agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:57]: At this step, if the runtime attachment is not done, the `agent.runtime.attached` property always has a `true` value. So, for consistency it is set to `false`. However, since the agent is not attached, the property value won't be used.... So, I will remove this piece of code. The code will be easier to read without apparently any impact.

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (jeanbisutti) [reply agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:57]: No, we have not to wait until the OTel PR is merged. OTel ```RuntimeAttach``` class exposes only one public method, ```attachJavaagentToCurrentJVM()```.

File: agent/runtime-attach/README.md
Comment (heyams) [inline agent/runtime-attach/README.md:4]: might want to use a placeholder for version. it will require an update each time a newer version of java jar is released.
something like {LATEST_JAVA_AGENT_VERSION}.jar will work.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java
Comment (jeanbisutti) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java:264]: Why? The problem is that the prefix has to contain only 4 characters? Another reason?

File: agent/runtime-attach/README.md
Comment (heyams) [inline agent/runtime-attach/README.md:1]: or alternatively, we can expose this runtime attach feature on azure doc and we don't need a separate readme for it?

image

under `Advanced configurations`?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java:264]: rp integration uses 4 chars. in your case only 3.

File: agent/runtime-attach/build.gradle.kts
Comment (trask) [inline agent/runtime-attach/build.gradle.kts:7]: similar to https://github.com/microsoft/ApplicationInsights-Java/blob/7eb023e32f5ec4987c09f618f326ff92baf41815/smoke-tests/apps/SpringBootAttachInMain/build.gradle.kts#L6

```suggestion
  implementation(project(":agent:agent", configuration = "shadow"))
```

File: agent/runtime-attach/build.gradle.kts
Comment (trask) [inline agent/runtime-attach/build.gradle.kts:8]: to avoid hard-coding this version, check out https://github.com/microsoft/ApplicationInsights-Java/blob/7eb023e32f5ec4987c09f618f326ff92baf41815/agent/instrumentation/applicationinsights-web-2.3/build.gradle.kts#L20-L25

File: agent/runtime-attach/build.gradle.kts
Comment (trask) [inline agent/runtime-attach/build.gradle.kts:6]: let's add `otelContribAlphaVersion` similar to `otelInstrumentationAlphaVersion`:

(we probably need to keep separate version constants for now since they are released from separate repos not always at the same time)

https://github.com/microsoft/ApplicationInsights-Java/blob/7eb023e32f5ec4987c09f618f326ff92baf41815/dependencyManagement/build.gradle.kts#L17

File: agent/runtime-attach/src/test/resources/applicationinsights.json
Comment (trask) [inline agent/runtime-attach/src/test/resources/applicationinsights.json:2]: just to be clear that nothing is listening there
```suggestion
  "connectionString": "InstrumentationKey=00000000-0000-0000-0000-0FEEDDADBEEF;IngestionEndpoint=http://dummy",
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java:272]:
```suggestion
    return Boolean.getBoolean("agent.runtime.attached");
```

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (trask) [inline agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:57]: do you know if we can rely on the classpath resource always having a real "file path"?

since we are just using the string value, do we need to convert `URL` -> `URI` -> `File` -> `String`? or can we use `URL.getFile()`?

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (trask) [inline agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:37]: it's helpful to scope any system properties, e.g. `applicationinsights.agent.runtime.attached`

and since this is only for internal (as opposed to customer) usage, maybe `applicationinsights.internal.runtime.attached`

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java
Comment (jeanbisutti) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/FirstEntryPoint.java]: Thank you for the explanations. I understand why now. The code is updated.

File: agent/runtime-attach/README.md
Comment (jeanbisutti) [reply agent/runtime-attach/README.md]: Good point @heyams! Done.

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (jeanbisutti) [reply agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java]: Good point @trask! Code updated

File: agent/runtime-attach/build.gradle.kts
Comment (jeanbisutti) [reply agent/runtime-attach/build.gradle.kts]: Code updated Please [recheck](https://github.com/microsoft/ApplicationInsights-Java/blob/runtime-attach/agent/runtime-attach/build.gradle.kts)

File: agent/runtime-attach/build.gradle.kts
Comment (jeanbisutti) [reply agent/runtime-attach/build.gradle.kts]: Good point. Code updated. Please check [this](https://github.com/microsoft/ApplicationInsights-Java/blob/runtime-attach/dependencyManagement/build.gradle.kts) and [this](https://github.com/microsoft/ApplicationInsights-Java/blob/runtime-attach/agent/runtime-attach/build.gradle.kts).

File: agent/runtime-attach/README.md
Comment (jeanbisutti) [reply agent/runtime-attach/README.md:1]: @trask What do you think? May I create a PR on Azure repo to report the Readme documentation now or later?

Comment (heyams) [reply agent/runtime-attach/README.md:1]: we can do that later when we're ready to make a new release.

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (heyams) [inline agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:78]:
```suggestion
    if (jsonUrl == null) {
```

File: agent/runtime-attach/src/test/resources/applicationinsights.json
Comment (heyams) [inline agent/runtime-attach/src/test/resources/applicationinsights.json:1]: without this file, it will return Optional.empty. does the test still work without json config?

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (trask) [reply agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:78]: this is a style choice, I would have chosen to inline this too, but I also don't mind being flexible when it comes to style choices that aren't covered by spotless, errorprone, or https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/style-guideline.md

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (trask) [reply agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:58]: > I have noticed [an environment variable related to the configuration content](https://github.com/microsoft/ApplicationInsights-Java/blob/7eb023e32f5ec4987c09f618f326ff92baf41815/agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/ConfigurationBuilder.java#L368). @trask I propose to add a system property doing the same thing.

this sounds good, it might be nice if we can hide it, maybe a static String in a class in the bootstrap classloader that we can populate from runtime attach and read from ConfigurationBuilder?

File: agent/runtime-attach/build.gradle.kts
Comment (trask) [inline agent/runtime-attach/build.gradle.kts:13]: `.8.2` looks wrong? probably this version is covered by dependency management (and probably assertj also?)

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (trask) [inline agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:35]: let's stay with OTel convention of lowercase `logger`

File: agent/runtime-attach/src/test/resources/applicationinsights.json
Comment (trask) [inline agent/runtime-attach/src/test/resources/applicationinsights.json:6]: the others aren't needed for this
```suggestion
  "connectionString": "InstrumentationKey=00000000-0000-0000-0000-0FEEDDADBEEF;IngestionEndpoint=http://dummy"
```

**Review by trask — APPROVED** [review]
👍

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (jeanbisutti) [reply agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:78]: My intention with the `jsonFileNotFound` variable was to give meaning to the `jsonUrl == null`  technical check.

File: agent/runtime-attach/src/test/resources/applicationinsights.json
Comment (jeanbisutti) [reply agent/runtime-attach/src/test/resources/applicationinsights.json:1]: The `RuntimeAttachTest` test will fail if we remove this file because _Application Insights_ needs a connection string provided in the json file. The connection string  may also be supplied with a system property or an environment variable, without a json file,  and the code execution path will then use the `Optional.empty()` line.

**Comment by jeanbisutti** [comment]
`RuntimeAttach` and `AgentFileProvider` classes could be removed once the 2 following OTel PRs will be merged and a new OTel release will be done:
* [OTel PR 1](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/354)
* [OTel PR 2](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/359)

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (trask) [inline agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:43]:
```suggestion
      "applicationinsights.internal.runtime.attach.json";
```

File: agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java
Comment (trask) [inline agent/runtime-attach/src/main/java/com/microsoft/applicationinsights/attach/ApplicationInsights.java:40]:
```suggestion
      "applicationinsights.internal.runtime.attached";
```

or maybe

```suggestion
      "applicationinsights.internal.runtime.attach.enabled";
```

**Review by trask — APPROVED** [review]
nice, a couple of suggestions on the internal property names if they make sense

**Pull request merged**
